File: functions/api/app/routers/organization.py
"""Organization settings router."""

import os
from uuid import UUID

from fastapi import APIRouter, HTTPException

from app.dependencies import CurrentOrg, CurrentUser, DBConnection
from app.schemas import DNSRecord, OrganizationResponse, OrganizationUpdate
from shared.ses_client import SESClient

SES_NOTIFICATION_TOPIC_ARN = os.environ.get("SES_NOTIFICATION_TOPIC_ARN", "")

router = APIRouter()


def format_dns_records(domain: str, tokens: list[str], region: str = "us-east-1") -> list[DNSRecord]:
    """Format all DNS records needed for SES domain verification.

    Returns DKIM CNAMEs, MAIL FROM (MX + SPF), and DMARC records.
    """
    records = []

    # DKIM CNAME records
    for token in tokens:
        records.append(
            DNSRecord(
                type="CNAME",
                name=f"{token}._domainkey.{domain}",
                value=f"{token}.dkim.amazonses.com",
            )
        )

    # MAIL FROM MX record (for custom bounce domain)
    mail_from_subdomain = f"mail.{domain}"
    records.append(
        DNSRecord(
            type="MX",
            name=mail_from_subdomain,
            value=f"10 feedback-smtp.{region}.amazonses.com",
        )
    )

    # MAIL FROM SPF record
    records.append(
        DNSRecord(
            type="TXT",
            name=mail_from_subdomain,
            value="v=spf1 include:amazonses.com ~all",
        )
    )

    # DMARC record
    records.append(
        DNSRecord(
            type="TXT",
            name=f"_dmarc.{domain}",
            value="v=DMARC1; p=none; rua=mailto:dmarc@{domain}".replace("{domain}", domain),
        )
    )

    return records


@router.get("", response_model=OrganizationResponse)
async def get_organization(
    org_id: CurrentOrg,
    db: DBConnection,
) -> OrganizationResponse:
    """Get current organization settings."""
    org = await db.fetchrow(
        """SELECT id, name, email_domain, email_domain_verified,
                  email_domain_dkim_tokens, email_from_name,
                  ses_tenant_name, ses_configuration_set
           FROM organizations WHERE id = $1""",
        org_id,
    )

    if not org:
        raise HTTPException(status_code=404, detail="Organization not found")

    dns_records = []
    if org["email_domain"] and org["email_domain_dkim_tokens"]:
        dns_records = format_dns_records(org["email_domain"], org["email_domain_dkim_tokens"])

    return OrganizationResponse(
        id=org["id"],
        name=org["name"],
        email_domain=org["email_domain"],
        email_domain_verified=org["email_domain_verified"] or False,
        email_from_name=org["email_from_name"],
        ses_tenant_name=org["ses_tenant_name"],
        ses_configuration_set=org["ses_configuration_set"],
        dns_records=dns_records,
    )


@router.patch("", response_model=OrganizationResponse)
async def update_organization(
    data: OrganizationUpdate,
    user: CurrentUser,
    org_id: CurrentOrg,
    db: DBConnection,
) -> OrganizationResponse:
    """Update organization settings including email domain."""
    # Check admin role for domain changes
    if data.email_domain is not None and user.get("role") != "admin":
        raise HTTPException(status_code=403, detail="Admin access required to change email domain")

    updates = []
    params = [org_id]
    param_idx = 2

    if data.email_from_name is not None:
        updates.append(f"email_from_name = ${param_idx}")
        params.append(data.email_from_name)
        param_idx += 1

    if data.email_domain is not None:
        if data.email_domain:
            # Start domain verification with SES
            ses = SESClient()
            result = ses.verify_domain(data.email_domain)

            if not result.get("success", True):
                raise HTTPException(
                    status_code=400,
                    detail=f"Failed to verify domain: {result.get('error_message', 'Unknown error')}",
                )

            updates.append(f"email_domain = ${param_idx}")
            params.append(data.email_domain)
            param_idx += 1

            updates.append(f"email_domain_dkim_tokens = ${param_idx}")
            params.append(result.get("dkim_tokens", []))
            param_idx += 1

            updates.append(f"email_domain_verified = ${param_idx}")
            params.append(result.get("verified", False))
            param_idx += 1
        else:
            # Clear domain
            updates.append(f"email_domain = ${param_idx}")
            params.append(None)
            param_idx += 1

            updates.append(f"email_domain_dkim_tokens = ${param_idx}")
            params.append(None)
            param_idx += 1

            updates.append(f"email_domain_verified = ${param_idx}")
            params.append(False)
            param_idx += 1

    if updates:
        await db.execute(
            f"""UPDATE organizations
                SET {', '.join(updates)}, updated_at = NOW()
                WHERE id = $1""",
            *params,
        )

    return await get_organization(org_id, db)


@router.post("/verify-domain", response_model=OrganizationResponse)
async def verify_domain_status(
    user: CurrentUser,
    org_id: CurrentOrg,
    db: DBConnection,
) -> OrganizationResponse:
    """Check and update domain verification status from SES."""
    if user.get("role") != "admin":
        raise HTTPException(status_code=403, detail="Admin access required")

    org = await db.fetchrow(
        "SELECT email_domain, ses_tenant_name, ses_configuration_set FROM organizations WHERE id = $1",
        org_id,
    )

    if not org or not org["email_domain"]:
        raise HTTPException(status_code=400, detail="No email domain configured")

    ses = SESClient()
    result = ses.get_domain_status(org["email_domain"])

    if not result.get("success", True):
        raise HTTPException(
            status_code=400,
            detail=f"Failed to check domain status: {result.get('error_message', 'Unknown error')}",
        )

    is_verified = result.get("verified", False)
    tenant_name = org["ses_tenant_name"]
    config_set_name = org["ses_configuration_set"]

    # Create SES tenant with configuration set when domain becomes verified
    if is_verified and not tenant_name and SES_NOTIFICATION_TOPIC_ARN:
        tenant_name = f"envoy-{org_id}"
        config_set_name = f"envoy-{org_id}"

        tenant_result = ses.setup_tenant(
            tenant_name=tenant_name,
            domain=org["email_domain"],
            configuration_set_name=config_set_name,
            sns_topic_arn=SES_NOTIFICATION_TOPIC_ARN,
        )
        if not tenant_result.get("success"):
            # Log but don't fail - domain verification is more important
            print(f"Failed to create SES tenant: {tenant_result.get('error_message')}")
            tenant_name = None
            config_set_name = None

    await db.execute(
        """UPDATE organizations
           SET email_domain_verified = $2,
               email_domain_dkim_tokens = $3,
               ses_tenant_name = COALESCE($4, ses_tenant_name),
               ses_configuration_set = COALESCE($5, ses_configuration_set),
               updated_at = NOW()
           WHERE id = $1""",
        org_id,
        is_verified,
        result.get("dkim_tokens", []),
        tenant_name,
        config_set_name,
    )

    return await get_organization(org_id, db)

File: lib/queries/organization.ts
import { sql } from "@/lib/db";
import { getEnv } from "@/lib/env";


type Row = Record<string, any>;

/** Columns allowed in dynamic UPDATE SET clauses — prevents SQL injection via key names. */
const ALLOWED_UPDATE_COLUMNS = new Set([
  "name", "email_domain", "email_domain_verified", "email_domain_dkim_tokens",
  "email_from_name", "ses_tenant_name", "ses_configuration_set",
  "agent_id", "environment_id", "vault_id",
]);

/** Resolved per-org Claude Managed Agents config. */
export interface AgentConfig {
  agentId: string;
  environmentId: string;
  /** Vault ids passed as session `vault_ids` so MCP servers can authenticate.
   * Empty when the org has no vault configured. */
  vaultIds: string[];
}

/**
 * Resolve an organization's Managed Agents config. `environment_id` falls back
 * to the deployment-wide `ANTHROPIC_DEFAULT_ENVIRONMENT_ID` when the org leaves
 * it blank. Returns null (→ treat as unconfigured, 503) when EITHER `agent_id`
 * OR the resolved `environment_id` is missing — `environment_id` is required by
 * `sessions.create`, so a configured-agent/no-env org must surface as
 * "not configured", never as a 502 on every call. Auth is the deployment-wide
 * `ANTHROPIC_API_KEY` (read by the SDK from env); there is no per-org key.
 */
export async function getAgentConfig(orgId: string): Promise<AgentConfig | null> {
  const rows = await sql`
    SELECT agent_id, environment_id, vault_id FROM organizations WHERE id = ${orgId}
  `;
  const row = rows[0];
  if (!row || !row.agent_id) return null;
  const orgEnv = row.environment_id;
  const environmentId =
    typeof orgEnv === "string" && orgEnv.length > 0
      ? orgEnv
      : getEnv().ANTHROPIC_DEFAULT_ENVIRONMENT_ID;
  if (!environmentId) return null;
  const orgVault = row.vault_id;
  const vaultIds =
    typeof orgVault === "string" && orgVault.length > 0 ? [orgVault] : [];
  return {
    agentId: String(row.agent_id),
    environmentId: String(environmentId),
    vaultIds,
  };
}

export async function getOrganization(orgId: string): Promise<Row | null> {
  // agent_id / environment_id are not secret (they only resolve to anything
  // with a valid ANTHROPIC_API_KEY), so unlike the old twin_api_key they are
  // selected and returned plainly.
  const rows = await sql`
    SELECT id, name, email_domain, email_domain_verified,
           email_domain_dkim_tokens, email_from_name,
           ses_tenant_name, ses_configuration_set,
           agent_id, environment_id, vault_id
    FROM organizations
    WHERE id = ${orgId}
  `;
  return rows[0] ?? null;
}

export async function updateOrganization(
  orgId: string,
  fields: Record<string, unknown>
): Promise<Row | null> {
  if (!fields || Object.keys(fields).length === 0) {
    return getOrganization(orgId);
  }

  const setClauses: string[] = [];
  const values: unknown[] = [];
  for (const [key, value] of Object.entries(fields)) {
    if (!ALLOWED_UPDATE_COLUMNS.has(key)) {
      throw new Error(`Unknown field: ${key}`);
    }
    values.push(value);
    setClauses.push(`${key} = $${values.length + 1}`);
  }

  if (setClauses.length === 0) {
    return getOrganization(orgId);
  }

  const query = `
    UPDATE organizations
    SET ${setClauses.join(", ")}, updated_at = NOW()
    WHERE id = $1
    RETURNING id, name, email_domain, email_domain_verified,
              email_domain_dkim_tokens, email_from_name,
              ses_tenant_name, ses_configuration_set,
              agent_id, environment_id, vault_id
  `;
  const rows = await sql.query(query, [orgId, ...values]);
  return rows[0] ?? null;
}

export async function getDomainVerificationStatus(
  orgId: string
): Promise<Row | null> {
  const rows = await sql`
    SELECT email_domain, email_domain_verified, email_domain_dkim_tokens,
           ses_tenant_name, ses_configuration_set
    FROM organizations
    WHERE id = ${orgId}
  `;
  return rows[0] ?? null;
}

export async function updateDomainVerification(
  orgId: string,
  verified: boolean,
  dkimTokens: string[],
  tenantName: string | null = null,
  configSetName: string | null = null
): Promise<void> {
  await sql`
    UPDATE organizations
    SET email_domain_verified = ${verified},
        email_domain_dkim_tokens = ${dkimTokens},
        ses_tenant_name = COALESCE(${tenantName}, ses_tenant_name),
        ses_configuration_set = COALESCE(${configSetName}, ses_configuration_set),
        updated_at = NOW()
    WHERE id = ${orgId}
  `;
}

/**
 * Format DNS records needed for SES domain verification.
 * Returns DKIM CNAMEs, MAIL FROM (MX + SPF), and DMARC records.
 */
export function formatDnsRecords(
  domain: string,
  tokens: string[],
  region: string = "us-east-1"
): Row[] {
  const records: Row[] = [];

  // DKIM CNAME records
  for (const token of tokens) {
    records.push({
      type: "CNAME",
      name: `${token}._domainkey.${domain}`,
      value: `${token}.dkim.amazonses.com`,
    });
  }

  // MAIL FROM MX record
  const mailFromSubdomain = `mail.${domain}`;
  records.push({
    type: "MX",
    name: mailFromSubdomain,
    value: `10 feedback-smtp.${region}.amazonses.com`,
  });

  // MAIL FROM SPF record
  records.push({
    type: "TXT",
    name: mailFromSubdomain,
    value: "v=spf1 include:amazonses.com ~all",
  });

  // DMARC record
  records.push({
    type: "TXT",
    name: `_dmarc.${domain}`,
    value: `v=DMARC1; p=none; rua=mailto:dmarc@${domain}`,
  });

  return records;
}
